### ISSUE TITLE
How to specify resources in various containers

### ISSUE THREAD

**Issue by elisiano** [coreos/kube-prometheus#376]
**What did you do?**
Similar to https://github.com/coreos/kube-prometheus/issues/372 I am in need of specifying resource requests/limits for all components. So far I failed at finding this info in the documentation and I got lost in the source code.
I need to be able to specify resources for the following containers (preferably via some jsonnet mixin):
```
 ➭ kubectl ops missing-resources
[...]
monitoring/alertmanager-main/alertmanager
monitoring/alertmanager-main/config-reloader
monitoring/prometheus-k8s/prometheus-config-reloader
monitoring/prometheus-k8s/rules-configmap-reloader
monitoring/prometheus-k8s/thanos-sidecar
[...]
```
(the above output is from a custom kubectl plugins which lists containers without requests).

Can anyone provide some guidance and/or reference to docs?

**Did you expect to see some different?**

**Environment**

* Prometheus Operator version:

    `quay.io/coreos/prometheus-operator:v0.34.0`
    

* Kubernetes version information:

    `v1.15.7`
    

* Kubernetes cluster kind:

    `kops`

**Comment by metalmatze** [comment]
All of these are coming from the Prometheus Operator.
In this case, you actually can't do anything with jsonnet, but need to adjust the Custom Resource for Prometheus and Alertmanager.

**Comment by elisiano** [comment]
@metalmatze can you point me to some example or at least the fields you're referring to?

**Comment by metalmatze** [comment]
Searching for `Resources` in here https://github.com/coreos/prometheus-operator/blob/master/pkg/apis/monitoring/v1/types.go might help.
Let me know how far that gets you. I am not totally sure about the resource for the sidecars.
It might be worth checking what we do in OpenShift: https://github.com/openshift/cluster-monitoring-operator/tree/master/assets

**Comment by elisiano** [comment]
Thank you @metalmatze this has been very useful, I managed to create a libsonnet file that I can import to setup the resources for most containers. I'm gonna paste it here in case someone in the future will find it useful:
```
{
  prometheus+:: {
    prometheus+: {
      spec+: {
        resources+: {
          limits+: {
            cpu: '2000m',
            memory: '8Gi'
          },
          requests+: {
            cpu: '200m',
            memory: '2Gi'
          },
        },
        thanos+: { // resources for the sidecar
          resources+: {
            limits+: {
              cpu: '4m',
              memory: '800Mi'
            },
            requests+: {
              cpu: '1m',
              memory: '20Mi'
            },
          }
        }
      }
    }
  },
  alertmanager+:: {
    alertmanager+: {
      spec+: {
        resources+: {
          limits:{
            cpu: '6m',
            memory: '200Mi'
          },
          requests:{
            cpu: '3m',
            memory: '20Mi'
          },
        }
      }
    }
  },
  // see https://github.com/coreos/kube-prometheus/blob/ce7c5fa3b45e5a613bcc4d99d3c6ba5ad38cef81/jsonnet/kube-prometheus/kube-prometheus-strip-limits.libsonnet#L20
  prometheusOperator+:: {
    deployment+: {
      spec+: {
        template+: {
          spec+: {
            local addArgs(c) =
              if c.name == 'prometheus-operator'
              then c + { args+: [ '--config-reloader-cpu=1m', '--config-reloader-memory=25Mi' ] }
              else c,
            containers: std.map(addArgs, super.containers),
          }
        }
      } 
    }
  }
}
```

That said there is one issue that I'm not sure I can solve on my own: the cpu/memory values set for the config-reloader are only for the limits. I couldn't find a way to specify the requests and looking at the  [code](https://github.com/coreos/prometheus-operator/blob/d01d6b094120576ede4d73a93e5e992f6b4c5e81/pkg/prometheus/statefulset.go#L733-L738) it would seem like only limits are set.

Should the requests be set at the same values as the limits or should there be other flags for the requests? (or is this left out intentionally for reasons that I couldn't figure out?)

**Comment by elisiano** [comment]
Closing this as it was addressed in #2958
